File: yggdrasil_decision_forests/cli/benchmark_inference.cc
/*
 * Copyright 2022 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Benchmarks the inference time of a model with the available inference
// engines.
//
// A model can be run with the "generic engine", a simple but slow
// implementation of the inference of the model that is always available and
// used as the golden implementation.
//
// Additional, "fast generic engines" are available with
// ListCompatibleFastEngines() or the "show_model" CLI. The availability of
// these engines depend on the structure of the model, the compile options and
// possibly the hardware. Note that engines should be linked to the binary.
//
// "fast generic engines" are accessible through a generic virtual interface, or
// by calling them directly. The second option being faster but require to know
// what interface is compatible at compilation time.
//
// The benchmark measures one copy of the dataset (from the best possible
// existing format; a simple memory copy in the best case) and one run of the
// engine.
//
// Usage example:
//
//   bazel run -c opt --copt=-mavx2 :benchmark_inference -- \
//     --alsologtostderr \
//     --model=/path/to/my/model \
//     --dataset=csv:/path/to/my/dataset.csv
//
// Result:
//
//   batch_size : 100  num_runs : 20
//   time/example(µs)  time/batch(µs)  method
//   ----------------------------------------
//   0.79025         79.025  GradientBoostedTreesQuickScorerExtended
//   9.179           917.9   GradientBoostedTreesGeneric
//   21.547          2154.8  Generic slow engine
//   ----------------------------------------
//
#include <string>
#include <vector>

#include "absl/flags/flag.h"
#include "absl/status/status.h"
#include "absl/time/time.h"
#include "absl/types/optional.h"
#include "yggdrasil_decision_forests/dataset/vertical_dataset.h"
#include "yggdrasil_decision_forests/dataset/vertical_dataset_io.h"
#include "yggdrasil_decision_forests/model/abstract_model.h"
#include "yggdrasil_decision_forests/model/model_library.h"
#include "yggdrasil_decision_forests/utils/benchmark/inference.h"
#include "yggdrasil_decision_forests/utils/logging.h"

ABSL_FLAG(std::string, model, "", "Path to model.");
ABSL_FLAG(std::string, dataset, "",
          "Typed path to dataset i.e. [type]:[path] format.");
ABSL_FLAG(
    int, num_runs, 20,
    "Number of times the dataset is run. Higher values increase the "
    "precision of the timings, but increase the duration of the benchmark.");
ABSL_FLAG(int, batch_size, 100,
          "Number of examples per batch. Note that some engine are not impacted"
          "by the batch size.");
ABSL_FLAG(int, warmup_runs, 1,
          "Number of runs through the dataset before the benchmark.");
ABSL_FLAG(bool, generic, true,
          "Evaluates the slow engine i.e. model->predict(). The "
          "generic engine is slow and mostly a reference. Disable it if the "
          "benchmark runs for too long.");

constexpr char kUsageMessage[] =
    "Benchmarks the inference time of a model with the available inference "
    "engines.";

namespace yggdrasil_decision_forests {

std::string ResultsToString(
    const utils::BenchmarkInferenceRunOptions& options,
    std::vector<utils::BenchmarkInferenceResult> results) {
  std::string report;

  // Sort the result from the fastest to the slowest.
  std::sort(results.begin(), results.end(), [](const auto& a, const auto& b) {
    return a.duration_per_example < b.duration_per_example;
  });

  absl::StrAppendFormat(&report, "batch_size : %d  num_runs : %d\n",
                        options.batch_size, options.runs->num_runs);
  absl::StrAppendFormat(&report, "time/example(us)  time/batch(us)  method\n");
  absl::StrAppendFormat(&report, "----------------------------------------\n");
  for (const auto& result : results) {
    absl::StrAppendFormat(
        &report, "%16.5g  %14.5g  %s\n",
        absl::ToDoubleMicroseconds(result.duration_per_example),
        absl::ToDoubleMicroseconds(result.duration_per_example *
                                   options.batch_size),
        result.name);
  }
  absl::StrAppendFormat(&report, "----------------------------------------\n");
  return report;
}

absl::Status Benchmark() {
  // Parse flags.
  const auto model_path = absl::GetFlag(FLAGS_model);
  if (model_path.empty()) {
    return absl::InvalidArgumentError("The --model is not specified.");
  }

  const auto dataset_path = absl::GetFlag(FLAGS_dataset);
  if (dataset_path.empty()) {
    return absl::InvalidArgumentError("The --dataset is not specified.");
  }
  const utils::BenchmarkInterfaceNumRunsOptions num_runs_options = {
      /*.num_runs =*/absl::GetFlag(FLAGS_num_runs),
      /*.warmup_runs =*/absl::GetFlag(FLAGS_warmup_runs),
  };
  const utils::BenchmarkInferenceRunOptions options{
      /*.batch_size =*/absl::GetFlag(FLAGS_batch_size),
      /*.runs =*/num_runs_options,
      /*.time =*/absl::nullopt};

  YDF_LOG(INFO) << "Loading model";
  std::unique_ptr<model::AbstractModel> model;
  RETURN_IF_ERROR(model::LoadModel(model_path, &model));
  YDF_LOG(INFO) << "The model is of type: " << model->name();

  YDF_LOG(INFO) << "Loading dataset";
  dataset::VerticalDataset dataset;
  RETURN_IF_ERROR(
      LoadVerticalDataset(dataset_path, model->data_spec(), &dataset,
                          /*ensure_non_missing=*/model->input_features()));

  std::vector<utils::BenchmarkInferenceResult> results;

  // Run engines.
  const auto engine_factories = model->ListCompatibleFastEngines();
  YDF_LOG(INFO) << "Found " << engine_factories.size()
                << " compatible fast engines.";
  for (const auto& engine_factory : engine_factories) {
    YDF_LOG(INFO) << "Running " << engine_factory->name();
    ASSIGN_OR_RETURN(auto engine, engine_factory->CreateEngine(model.get()));
    RETURN_IF_ERROR(utils::BenchmarkFastEngine(options, *engine.get(),
                                               *model.get(), dataset, &results,
                                               engine_factory->name()));
  }

  if (absl::GetFlag(FLAGS_generic)) {
    YDF_LOG(INFO) << "Running the slow generic engine";
    RETURN_IF_ERROR(
        utils::BenchmarkGenericSlowEngine(options, *model, dataset, &results));
  }

  // Show results.
  std::cout << ResultsToString(options, results);
  return absl::OkStatus();
}

}  // namespace yggdrasil_decision_forests

int main(int argc, char** argv) {
  InitLogging(kUsageMessage, &argc, &argv, true);
  const auto status = yggdrasil_decision_forests::Benchmark();
  if (!status.ok()) {
    YDF_LOG(INFO) << "The benchmark failed with the following error: "
                  << status;
    return 1;
  }
  return 0;
}

File: yggdrasil_decision_forests/port/go/cli/benchmark_inference/benchmark_inference.go
/*
 * Copyright 2022 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
Benchmark the inference speed of a model.

Usage example:

	# Disable CPU power scaling
	sudo apt install linux-cpupower
	sudo cpupower frequency-set --governor performance

	# Configure benchmark
	MODEL=$(pwd)/learning/lib/ami/simple_ml/test_data/model/adult_gdt_classifier
	DATASET=csv:$(pwd)/learning/lib/ami/simple_ml/test_data/dataset/adult.csv
	BUILD_FLAGS="-c opt --copt=-mfma --copt=-mavx2"
	RUN_OPTIONS="--model=${MODEL} \
		--dataset=${DATASET} \
		--batch_size=100 \
		--warmup_runs=10 \
		--num_runs=100"

	# Benchmark
	bazel run ${BUILD_FLAGS} \
		//third_party/yggdrasil_decision_forests/port/go/cli:benchmarkinference -- \
		${RUN_OPTIONS}

Naming convention:
  - A (benchmark) "run" evaluates the speed of a model on a dataset.
  - A "run" is composed of one of more "unit runs".
  - A "unit run" measure the speed of a specific inference implementation (called "engine") with
    specific parameters (e.g. batchSize=10).
*/
package main

import (
	"context"
	"encoding/csv"
	"fmt"
	"strings"
	"time"

	"flag"
	model_io "github.com/google/yggdrasil-decision-forests/yggdrasil_decision_forests/port/go/model/io/canonical"
	"github.com/google/yggdrasil-decision-forests/yggdrasil_decision_forests/port/go/serving/engine"
	"github.com/google/yggdrasil-decision-forests/yggdrasil_decision_forests/port/go/serving/example"
	"github.com/google/yggdrasil-decision-forests/yggdrasil_decision_forests/port/go/serving"

	// External dependencies, pls keep in this position in file.
	"github.com/google/yggdrasil-decision-forests/yggdrasil_decision_forests/port/go/utils/file"
	// End of external dependencies.//
	//
)

var flagModel = flag.String("model", "", "Path to the model")
var flagDataset = flag.String("dataset", "", "Type path to the dataset e.g. csv:/tmp/my_file.csv")
var flagNumRuns = flag.Int("num_runs", 20, "Number of times the dataset is run. Higher values increase the precision of the timings, but increase the duration of the benchmark.")
var flagBatchSize = flag.Int("batch_size", 100, "Number of examples per batch. Note that some engine are not impacted by the batch size.")
var flagWarmupRuns = flag.Int("warmup_runs", 2, "Number of runs through the dataset before the benchmark.")

// Options are the options to run the benchmark.
type Options struct {
	// Number of times the entire dataset is run.
	numRuns int

	// Number of runs to "warmup" the engine i.e. running the engine before the benchmark.
	warmupRuns int

	// Number of examples in each batch. Some engine speed can be impacted by the batch size.
	batchSize int
}

// Run runs the benchmark. The results are printed on the standard output.
func Run(modelPath string, datasetPath string, options *Options) error {
	fmt.Printf("Run benchmark with\n  model: %v\n  dataset: %v\n  options: %v\n",
		modelPath, datasetPath, options)

	// Check the validity of the options
	if options.numRuns <= 0 {
		return fmt.Errorf("options.runs should be greater or equal to 1")
	}
	if options.batchSize <= 0 {
		return fmt.Errorf("options.batchSize should be greater or equal to 1")
	}
	if options.warmupRuns <= 0 {
		return fmt.Errorf("options.warmupRuns should be greater or equal to 1")
	}

	// Load the model
	fmt.Println("Load model")
	model, err := model_io.LoadModel(modelPath)
	if err != nil {
		return err
	}
	fmt.Println("\tFound model:", model.Name())

	// Compile the model
	fmt.Println("Compile model")
	engine, err := serving.NewEngine(model)
	if err != nil {
		return err
	}
	fmt.Printf("\tBuilt engine \"%T\" with %d input features\n",
		engine, engine.Features().NumFeatures())

	// Loads the dataset.
	fmt.Println("Load dataset")
	dataset, err := loadDataset(engine, datasetPath)
	if err != nil {
		return err
	}
	fmt.Printf("\t%d examples\n", dataset.NumAllocatedExamples())

	// Run the benchmark
	fmt.Println("Run benchmark")
	// For now, the benchmark is composed of a single evaluation.
	result, err := UnitRun(engine, dataset, options)
	if err != nil {
		return err
	}

	// Print the result
	fmt.Println("Results")
	fmt.Print(result)

	return nil
}

func (result *UnitRunResult) String() string {
	return fmt.Sprintf(
		`Avg. time per dataset:  %v
Avg. time per batch:    %v
Avg. time per examples: %v
`,
		// Note: In Go, duration * duration gives a duration, where the result is effectively
		// numNanoseconds * numNanoseconds -> numNanoseconds.
		result.durationPerExample*time.Duration(result.numExamples),
		result.durationPerExample*time.Duration(result.batchSize),
		result.durationPerExample)
}

func loadDataset(engine engine.Engine, typedPath string) (*example.Batch, error) {
	format, path, err := parseTypedPath(typedPath)
	if err != nil {
		return nil, err
	}
	switch format {
	case "csv":
		return loadDatasetCsv(engine, path)
	default:
		return nil, fmt.Errorf("Non supported dataset format %v", format)
	}
}

// parseTypedPath parses a typed path into its constituents.
//
// For example:
//
//	Input: "csv:/path/to/sharded/csv/file@10"
//	Results:
//	  1. "csv"
//	  2. "/path/to/sharded/csv/file@10"
//	  3. nil (i.e. no error)
func parseTypedPath(typedPath string) (pathType string, path string, err error) {
	i := strings.Index(typedPath, ":")
	if i == -1 {
		err = fmt.Errorf("Malformed typed dataset path. Expecting [format]:[path]. Instead, got %v", typedPath)
		return
	}
	pathType = typedPath[:i]
	path = typedPath[i+1:]
	err = nil
	return
}

func loadDatasetCsv(engine engine.Engine, path string) (*example.Batch, error) {
	// Read the csv content.
	fileHandle, err := file.OpenRead(context.Background(), path)
	if err != nil {
		return nil, err
	}
	fileIO := fileHandle.IO(context.Background())
	defer fileIO.Close()
	csvData, err := csv.NewReader(fileIO).ReadAll()
	if err != nil {
		return nil, err
	}

	// Skip the header file
	csvHeader := csvData[0]
	csvData = csvData[1:]
	numExamples := len(csvData)

	// Convert the csv into a "Batch".
	examples := engine.AllocateExamples(numExamples)
	for exampleIdx := 0; exampleIdx < numExamples; exampleIdx++ {
		if err := examples.SetFromFields(exampleIdx, csvHeader, csvData[exampleIdx]); err != nil {
			return nil, err
		}
	}
	return examples, nil
}

// UnitRunResult contains the benchmark result for a single run.
type UnitRunResult struct {
	durationPerExample time.Duration
	numExamples        int
	batchSize          int
}

// UnitRun benchmark a single engine on a give dataset.
func UnitRun(engine engine.Engine, dataset *example.Batch, options *Options) (*UnitRunResult, error) {

	batchSize := options.batchSize
	numExamples := dataset.NumAllocatedExamples()
	numBatches := (numExamples + batchSize - 1) / options.batchSize

	batch := engine.AllocateExamples(batchSize)
	predictions := engine.AllocatePredictions(batchSize)

	run := func(numRuns int) {
		for runIdx := 0; runIdx < numRuns; runIdx++ {
			for batchIdx := 0; batchIdx < numBatches; batchIdx++ {
				beginIdx := batchIdx * batchSize
				endIdx := (batchIdx + 1) * batchSize
				if endIdx > numExamples {
					endIdx = numExamples
				}
				numExamplesInBatch := endIdx - beginIdx

				// Set the example values.
				// The benchmark time account for a single copy of the feature values.
				batch.CopyFrom(dataset, beginIdx, endIdx)

				// Generate the predictions.
				engine.Predict(batch, numExamplesInBatch, predictions)
			}
		}
	}

	// Warmup
	_ = time.Now()
	run(options.warmupRuns)
	_ = time.Now()

	// Benchmark
	start := time.Now()
	run(options.numRuns)
	end := time.Now()

	result := &UnitRunResult{
		numExamples: numExamples,
		batchSize:   batchSize,
	}
	result.durationPerExample = end.Sub(start) / time.Duration(options.numRuns*numExamples)
	return result, nil
}

func main() {
	flag.Parse()

	options := Options{
		numRuns:    *flagNumRuns,
		batchSize:  *flagBatchSize,
		warmupRuns: *flagWarmupRuns}
	Run(*flagModel, *flagDataset, &options)
}
